engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java

```diff
@@ -15,44 +15,68 @@
  */
 package org.terasology.rendering.dag;
 
+import com.google.common.collect.HashMultimap;
 import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
 import org.terasology.engine.SimpleUri;
 
+import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 /**
  * TODO: Add javadocs
  */
 public class RenderGraph { // TODO: add extends DirectedAcyclicGraph<Node>
-    private List<Node> nodes;
-
-    public RenderGraph() {
-        nodes = Lists.newArrayList();
-    }
+    private Map<SimpleUri, Node> nodes = Maps.newHashMap();
+    private Multimap<Node, Node> nodeEdges = HashMultimap.create();
 
     public SimpleUri addNode(Node node, String suggestedUri) {
-        nodes.add(node);
-
         // TODO: make sure URIs are actually unique: if "myModule:blur" is present the node gets the uri "myModule:blur2" instead.
         // TODO: make sure the namespace in the uri is engine-assigned, so that only engine nodes can have the "engine:" namespace - everything else gets the namespace of the module.
         SimpleUri nodeUri = new SimpleUri("engine:" + suggestedUri);
-        node.setUri(nodeUri);
+
+        nodes.put(nodeUri, node);
```

Comment:
I've been thinking about this, and although I agree with the general idea of your approach, I don't want every single node uri to end with 0. I really like `dagNodeCommand engine:outputToScreenNode ...` instead of `dagNodeCommand engine:outputToScreenNode0 ...` .
So what I propose is, we do exactly what you said, minus the last line.
So, the node naming would go something like:
myNode, myNode2, myNode3 and so on.

Reply:
I thought about it and I would prefer that the first node either ends with 0 (programming-centric) or 1 (user-centric). 

I.e. thinking in the context of the VDAG, non-programming users would wonder why there is a myNode without number and then the count jumps to myNode2. On the other hand it would be also confusing to have myNode and then myNode1 - which one is the first, really? Having the count start from the very first node with a number would make everything more homogeneous. 

That been said, this isn't a blocker for me and if you really want to you can go ahead without a number for the first node.

Reply:
Actually, I never mentioned myNode1 - I jumped right from myNode to myNode2.

The only reason I'm thinking along this line is, there will be very few collisions to begin with. So, the user might get confused why there is a 1 after every node, when there are no duplicates in the first palce.